### PR TITLE
#issue-9: Filter the shopping Lists

### LINKED ISSUE #9: 9. As a user, I want to filter my shopping list to make it easier to locate an item in the list.
AC:

* Display a text field above the top of the shopping list
* As the user types into the field, the list should narrow to display only items that contain the text the user entered in the filter field
* When the field has text in it, the user should be able to tap a UI element (e.g., with an "X" button next to the field) to clear the field
* The filter text should match any part of the item name (i.e. it should _not_ only match from the start of the string)

### PR DESCRIPTION
**Pull request by hajarNasr**
## Type of Changes

|      | Type                       |
| --- | -------------------------- |
|  ✓    | :sparkles: New feature     |
|  ✓    |  🔨  Refactoring  |

## Description
**This PR adds new functionality to the app and refactors some of the previous code**

**First**: It adds a text filed and a clear-text-field button in `List.js` to allow users to enter or then clear search terms.
**Second**: It declares the `getItemsKeys(shoppingList, searchTerm)` function in `List.js` which takes a `shoppingList` and `searchTerm` and returns the items' keys according to that data. 
**Third**: It creates the `SortedList` component for displaying the items on the user's shopping list and refactors a few other things.

## Related Issue
Closes #9 

## Acceptance Criteria
- [X] Display a text field above the top of the shopping list
- [X] As the user types into the field, the list should narrow to display only items that contain the text the user entered in the filter field
- [X] When the field has text in it, the user should be able to tap a UI element (e.g., with an "X" button next to the field) to clear the field
- [X] The filter text should match any part of the item name (i.e. it should not only match from the start of the string)

## Update
Users can filter their shopping lists by entering search terms into the text field added in the `List` route, and the list will only display the items whose names include these search terms. *If nothing matches, nothing shows*.

## How things look so far
a gif of how users use the app to search for items

## Testing Steps / QA Criteria

* From your terminal, pull down this branch with `git pull origin hn-dg-issue-9` 
* Check that branch out with `git checkout hn-dg-issue-9`
* Run `yarn start` to run the [local server](http://localhost:3000/).
* Join a list by using the **bushy ret morale** token, or create your own list and add some items to it.
* If you're not already, go to the `List` route and use the text field to search items on the list.

Now, you **can** filter the shopping list by anything you enter into the text field. If there are no items whose names include your search term, nothing should show up.

Thank you!

### CODE REVIEW

File: src/components/List.js [hunk 1]
```diff
@@ -1,39 +1,42 @@
...
-                        <ListItem
-                          listItem={data[key]}
-                          key={key}
-                          listId={data.id}
-                          itemId={key}
-                        />
-                      );
-                    })}
-                </ul>
+              {itemsKeys.length || searchTerm ? (
```
Comment (saraheisa) [inline src/components/List.js:38]: Good Job here 👍
you can also add another check if `searchTerm && itemsKeys.length === 0` to view a message to the user if non of the items matches the search term

File: src/components/List.js [hunk 2]
```diff
@@ -1,39 +1,42 @@
...
+                <>
+                  <div className="search-term-wrapper">
+                    <input
+                      value={searchTerm}
+                      onChange={(e) => setSearchTerm(e.target.value.trim())}
```
Comment (saraheisa) [inline src/components/List.js]: I think no need to call `trim()` here because it's already done in the `filter `function (`removePunctuation`)

File: src/components/List.js [hunk 3]
```diff
@@ -1,39 +1,42 @@
...
+                      placeholder="Search item"
+                    />
+                    <button
+                      onClick={() => setSearchTerm('')}
+                      disabled={searchTerm ? false : true}
```
Comment (saraheisa) [inline src/components/List.js:48]: I would prefer to use `hidden` here instead of `disabled` just a personal preference :D

File: src/components/List.js [hunk 4]
```diff
@@ -44,3 +47,16 @@ export default function List() {
     </div>
   );
 }
+
+const getItemsKeys = (shoppingList, searchTerm) => {
```
Comment (saraheisa) [inline src/components/List.js]: I like how you contained the search logic in a separate function and using `removePunctuation` before searching is def a good idea Well Done 👍
Maybe later we can also highlight the `searchTerm` in the items while searching [like here](https://bitsofco.de/a-one-line-solution-to-highlighting-search-matches) will be a fancy addition to the app 💅

File: src/components/SortedList.js [hunk 5]
```diff
@@ -0,0 +1,17 @@
+import React from 'react';
+import ListItem from './ListItem';
+
+const SortedList = ({ itemsKeys, data }) => {
```
Comment (saraheisa) [inline src/components/SortedList.js:4]: 👍👌

**Review by saraheisa — APPROVED** [review]
Everything is working as expected and neatly done 👌
Good job as usual 🙌

File: src/components/List.js (hunk 1 above)
Comment (hajarNasr) [reply src/components/List.js:38]: I think that's a good idea! Thank you, Sarah! :)

File: src/components/List.js (hunk 2 above)
Comment (hajarNasr) [reply src/components/List.js]: We actually had that problem with the text input and was fixed when we added `trim()`. I can't remember it now tbh but I'll try to look at the code again tomorrow to remember why we added `trim()` here. 😅

File: src/components/List.js (hunk 3 above)
Comment (hajarNasr) [reply src/components/List.js:48]: We were actually using it with something like `{searchTerm && theBtn}` but it felt a bit jumpy, like in the gif (yes, we were too lazy to make a new gif after changing the button 😅). Anyway, we have plenty of time for styling in later issues. We can change EVERYTHING then. I've already started to dislike how the app looks haha. 😸

File: src/components/List.js (hunk 4 above)
Comment (hajarNasr) [reply src/components/List.js]: Yes, I think highlighting the search term is a great idea. 👍 
Thank you so much for this good review, Sarah. ❤️

File: src/components/SortedList.js [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+  return (
+    <ul>
+      {itemsKeys
+        // a stolen line from Andrew's blog
```
Comment (yahayakenny) [inline src/components/SortedList.js]: Hahaha, Nice! i guess we have to thank Andrew for giving us an efficient method to sort. Great work implementing it here

File: src/components/List.js (hunk 1 above)
Comment (yahayakenny) [reply src/components/List.js:38]: Great suggestion here Sarah.

File: src/components/List.js (hunk 2 above)
Comment (deboragaleano) [reply src/components/List.js]: Thanks Sarah, I think you're right. We  probably added it _before_ we imported the `filter` function.

File: debug.log [hunk 7]
```diff
@@ -0,0 +1 @@
+[1103/185007.341:ERROR:directory_reader_win.cc(43)] FindFirstFile: The system cannot find the path specified. (0x3)
```
Comment (segdeha) [inline debug.log:1]: What’s this file about? Maybe we should add `*.log` to our `.gitignore`?

File: src/components/AddItem.js [hunk 8]
```diff
@@ -30,11 +29,10 @@ const AddItem = () => {
     const newItem = {
       name: inputValue,
       lastPurchased: null,
-      howSoon: e.target['how-soon'].value,
+      howSoon: parseInt(e.target['how-soon'].value),
```
Comment (segdeha) [inline src/components/AddItem.js]: Curious what prompted adding the `parseInt` here?

When using `parseInt`, it’s good form to pass a second parameter for the base. The default is base 10, which is probably what you intend, but by passing it, you let the next person who comes across it know that it was purposefully set to that base.

Also, if you’re just trying to convert a number that’s a string into a proper `Number` (as opposed to trying to extract a number from a string with other characters such as `800px`), you could use the `+` operator, as follows:

    howSoon: +e.target['how-soon'].value,

File: src/components/List.js [hunk 9]
```diff
@@ -1,39 +1,46 @@
...
+import { filter as removePunctuation } from '../lib/helper';
+import '../css/components/ItemsList.css';
 
 export default function List() {
+  const [searchTerm, setSearchTerm] = useState('');
+
   return (
     <div>
-      <h1>LIST OF ITEMS</h1>
+      <h3>LIST OF ITEMS</h3>
```
Comment (segdeha) [inline src/components/List.js]: Curious why the change from `<h1>` to `<h3>`? If it’s just to get a header that is styled more appropriately, that’s better done in CSS. `<h1>` is _probably_ the right tag, semantically, especially considering your header is in a `<div>`

File: src/components/List.js [hunk 10]
```diff
@@ -1,39 +1,46 @@
...
       <FirestoreDocument
         path={`shoppingLists/${getToken()}`}
         render={({ isLoading, data }) => {
-          const itemsKeys = data
-            ? Object.keys(data).filter((key) => key !== 'id')
-            : '';
+          let itemsKeys = data ? getItemsKeys(data, searchTerm) : [];
```
Comment (segdeha) [inline src/components/List.js:33]: This line was a little hard for me to parse at first. Consider leaving a comment explaining what it’s doing. Thanks!

File: src/components/SortedList.js (hunk 6 above)
Comment (segdeha) [reply src/components/SortedList.js]: 👏👏👏

File: src/components/SortedList.js [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+        .sort((a, b) => (a > b ? 1 : a < b ? -1 : 0))
```
Comment (segdeha) [inline src/components/SortedList.js:11]: Of course, I understand this because it comes from my blog post, but it might be worth leaving a comment here for future you to be able to decipher it 😅

File: src/lib/helper.js [hunk 12]
```diff
@@ -21,4 +21,4 @@ const getItemsNamesFromDoc = (items) => {
 const fromMilleToHours = (time) =>
   (Date.now() - new Date(time)) / (1000 * 60 * 60);
 
-export { existingName, fromMilleToHours };
+export { existingName, fromMilleToHours, filter };
```
Comment (segdeha) [inline src/lib/helper.js:24]: Smart re-use of the existing `filter` function!

**Review by segdeha — APPROVED** [review]
I left a couple of questions and suggestions, but nothing blocking. Works great, nice job!

File: src/components/AddItem.js (hunk 8 above)
Comment (hajarNasr) [reply src/components/AddItem.js]: Yes! We were trying to convert the value into `int` before storing it in the database.
Thank you for letting us know about the base 10 and the `+` operator. :)

File: debug.log (hunk 7 above)
Comment (hajarNasr) [reply debug.log:1]: That file was actually on my local branch and I tried to delete it but couldn't.  
Yes, sure! Will definitely add it to `.gitignore`.

File: src/components/List.js (hunk 10 above)
Comment (deboragaleano) [reply src/components/List.js:33]: Yes for sure! :)

File: src/components/AddItem.js (hunk 8 above)
Comment (deboragaleano) [reply src/components/AddItem.js]: This is actually new to me (the + operator), thanks so much Andrew!

File: src/components/List.js (hunk 9 above)
Comment (hajarNasr) [reply src/components/List.js]: Yes! I thought it was too big. Sorry, my mistake. 😅 
I agree, I should've done it in CSS. 👍

File: src/components/SortedList.js [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+import ListItem from './ListItem';
+
+const SortedList = ({ itemsKeys, data }) => {
+  return (
+    <ul>
+      {itemsKeys
+        // This explicitly compares between a or b, in order to sort the items by the keys
+        // If a is greater than b, it will return 1, else if a is less than b it will return -1
+        // Or else, it will return 0
+        .sort((a, b) => (a > b ? 1 : a < b ? -1 : 0))
```
Comment (Lukabaramishvili) [inline src/components/SortedList.js:11]: These comments are super helpful!

**Review by Lukabaramishvili — APPROVED** [review]
Great work here. Everything works as expected! 👍
